File: src/llm_pricer.rs
//! WAIFU L1 - The Oracle of Truth
//! LLM-based pricing engine that replaces AMMs and order books

use crate::types::{Transaction, TransactionContext, Operation, Priority, SovereignAsset, AssetType};
use crate::state::AgenticState;
use candle_core::{Device, Tensor, DType, Result as CandleResult};
use std::sync::Arc;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum PricerError {
    #[error("Model loading failed: {0}")]
    ModelLoadError(String),
    #[error("Inference failed: {0}")]
    InferenceError(String),
    #[error("Invalid context: {0}")]
    InvalidContext(String),
    #[error("Candle error: {0}")]
    CandleError(#[from] candle_core::Error),
}

/// The deterministic LLM pricing engine
/// Temperature = 0.0 for consensus-compatible inference
pub struct LlmPricer {
    /// Device for tensor operations (CPU/GPU)
    device: Device,
    /// Model weights hash (for PoI proof)
    pub model_hash: [u8; 32],
    /// Embedding dimension
    embed_dim: usize,
    /// Context weights for pricing factors
    context_weights: Tensor,
    /// Operation-specific pricing matrices
    operation_matrices: Vec<Tensor>,
    /// Temperature (must be 0.0 for determinism)
    temperature: f64,
}

impl LlmPricer {
    /// Load a quantized model from safetensors format
    pub fn load_quantized(model_path: &str) -> Result<Self, PricerError> {
        // In production, this loads actual quantized transformer weights
        // For now, we initialize with deterministic pseudo-random weights
        let device = Device::Cpu;
        let embed_dim = 768;
        
        // Generate deterministic weights based on model path hash
        let model_hash = *blake3::hash(model_path.as_bytes()).as_bytes();
        
        // Initialize context weighting tensor
        let context_weights = Tensor::ones((embed_dim, 64), DType::F32, &device)?;
        
        // Initialize operation-specific matrices (one per operation type)
        let operation_matrices: Vec<Tensor> = (0..8)
            .map(|i| {
                let seed_tensor = Tensor::ones((embed_dim, embed_dim), DType::F32, &device)
                    .unwrap();
                // Scale each matrix differently for operation-specific behavior
                (&seed_tensor * (1.0 / (i as f64 + 1.0))).unwrap()
            })
            .collect();
        
        Ok(Self {
            device,
            model_hash,
            embed_dim,
            context_weights,
            operation_matrices,
            temperature: 0.0, // CRITICAL: Must be 0 for consensus
        })
    }

    /// Price a transaction based on its context and current network state
    /// Returns (clearing_rate, pricing_rationale)
    pub fn price_transaction(
        &self,
        tx: &Transaction,
        state: &AgenticState,
    ) -> Result<(f64, String), PricerError> {
        // 1. Encode the transaction context into a tensor
        let context_embedding = self.encode_context(&tx.context)?;
        
        // 2. Get the operation-specific pricing matrix
        let op_matrix = self.get_operation_matrix(&tx.context.operation);
        
        // 3. Query network state for supply/demand signals
        let network_signal = self.compute_network_signal(state, &tx.from, &tx.to);
        
        // 4. Perform deterministic matrix multiplication (THE CORE PRICING)
        let price_tensor = context_embedding.matmul(&op_matrix)?;
        let price_vector = price_tensor.matmul(&self.context_weights)?;
        
        // 5. Reduce to scalar clearing rate
        let raw_price: f64 = price_vector.sum_all()?.to_scalar()?;
        
        // 6. Apply network signal modulation
        let clearing_rate = self.normalize_price(raw_price, network_signal, &tx.context);
        
        // 7. Generate pricing rationale (deterministic)
        let rationale = self.generate_rationale(&tx.context, clearing_rate, network_signal);
        
        Ok((clearing_rate, rationale))
    }

    /// Encode transaction context into embedding space
    fn encode_context(&self, context: &TransactionContext) -> CandleResult<Tensor> {
        // Create a feature vector from context properties
        let mut features = vec![0.0f32; self.embed_dim];
        
        // Energy budget encoding
        features[0] = context.energy_budget as f32;
        
        // Priority encoding (one-hot style)
        let priority_idx = match context.priority {
            Priority::Low => 1,
            Priority::Normal => 2,
            Priority::High => 3,
            Priority::Critical => 4,
        };
        features[priority_idx] = 1.0;
        
        // Operation type encoding
        let op_idx = match &context.operation {
            Operation::Transfer { amount } => {
                features[10] = *amount as f32;
                10
            }
            Operation::Execute { .. } => 20,
            Operation::Deploy { initial_energy, .. } => {
                features[30] = *initial_energy as f32;
                30
            }
            Operation::Bridge { .. } => 40,
            Operation::Stake { amount } => {
                features[50] = *amount as f32;
                50
            }
            Operation::Infer { max_tokens, .. } => {
                features[60] = *max_tokens as f32;
                60
            }
            Operation::Swap { .. } => 70,
        };
        features[op_idx] = 1.0;
        
        // Payload size factor
        features[100] = context.payload.len() as f32 / 1024.0;
        
        // Oracle refs count
        features[101] = context.oracle_refs.len() as f32;
        
        Tensor::from_vec(features, (1, self.embed_dim), &self.device)
    }

    /// Get the appropriate pricing matrix for an operation type
    fn get_operation_matrix(&self, operation: &Operation) -> &Tensor {
        let idx = match operation {
            Operation::Transfer { .. } => 0,
            Operation::Execute { .. } => 1,
            Operation::Deploy { .. } => 2,
            Operation::Bridge { .. } => 3,
            Operation::Stake { .. } => 4,
            Operation::Infer { .. } => 5,
            Operation::Swap { .. } => 6,
        };
        &self.operation_matrices[idx.min(self.operation_matrices.len() - 1)]
    }

    /// Compute network-level supply/demand signal
    fn compute_network_signal(
        &self,
        state: &AgenticState,
        from: &crate::types::AgentId,
        to: &crate::types::AgentId,
    ) -> f64 {
        // Get sender's reputation/stake
        let sender_stake = state.get_agent_stake(from);
        let receiver_stake = state.get_agent_stake(to);
        
        // Get network congestion
        let congestion = state.get_network_congestion();
        
        // Get total network compute availability
        let total_compute = state.get_total_compute();
        
        // Combine signals (deterministic formula)
        let reputation_factor = (sender_stake + receiver_stake).ln_1p();
        let congestion_factor = 1.0 + (congestion * 2.0);
        let compute_factor = (total_compute / 1_000_000.0).sqrt();
        
        reputation_factor * congestion_factor / compute_factor.max(1.0)
    }

    /// Normalize raw price to usable clearing rate
    fn normalize_price(&self, raw: f64, network_signal: f64, context: &TransactionContext) -> f64 {
        // Base normalization
        let base = raw.abs() / (self.embed_dim as f64 * 64.0);
        
        // Apply network signal
        let signaled = base * (1.0 + network_signal);
        
        // Apply priority multiplier
        let priority_mult = match context.priority {
            Priority::Low => 0.8,
            Priority::Normal => 1.0,
            Priority::High => 1.5,
            Priority::Critical => 2.5,
        };
        
        // Final clearing rate (bounded)
        (signaled * priority_mult).max(0.0001).min(1_000_000.0)
    }

    /// Generate deterministic pricing rationale
    fn generate_rationale(
        &self,
        context: &TransactionContext,
        clearing_rate: f64,
        network_signal: f64,
    ) -> String {
        let op_name = match &context.operation {
            Operation::Transfer { amount } => format!("TRANSFER({:.2})", amount),
            Operation::Execute { function, .. } => format!("EXECUTE({})", function),
            Operation::Deploy { .. } => "DEPLOY".to_string(),
            Operation::Bridge { source_chain, .. } => format!("BRIDGE({})", source_chain),
            Operation::Stake { amount } => format!("STAKE({:.2})", amount),
            Operation::Infer { max_tokens, .. } => format!("INFER({})", max_tokens),
            Operation::Swap { .. } => "SWAP".to_string(),
        };
        
        format!(
            "PoI-PRICE: {} @ {:.6} | NET_SIG: {:.4} | PRI: {:?} | ENERGY: {:.2}",
            op_name, clearing_rate, network_signal, context.priority, context.energy_budget
        )
    }

    /// Assess the utility value of a legacy asset being bridged
    pub fn assess_utility_vector(&self, legacy_value: f64) -> f64 {
        // AI utility tokens get premium, memecoins get discounted
        // In production, this queries the agent's historical performance
        
        // Base conversion rate
        let base_rate = 0.95;
        
        // Value-based scaling (larger values = slightly better rate due to network effects)
        let scale_factor = (legacy_value.ln_1p() / 10.0).min(1.2);
        
        base_rate * scale_factor
    }

    /// Batch price multiple transactions for parallel execution
    pub fn price_batch(
        &self,
        transactions: &[Transaction],
        state: &AgenticState,
    ) -> Vec<Result<(f64, String), PricerError>> {
        // Use rayon for parallel pricing
        use rayon::prelude::*;
        
        transactions
            .par_iter()
            .map(|tx| self.price_transaction(tx, state))
            .collect()
    }

    /// Get the model hash for PoI proof
    pub fn get_model_hash(&self) -> [u8; 32] {
        self.model_hash
    }

    /// Verify that inference is deterministic (temperature = 0)
    pub fn is_deterministic(&self) -> bool {
        self.temperature == 0.0
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::AgentId;

    #[test]
    fn test_pricer_determinism() {
        let pricer = LlmPricer::load_quantized("test-model.safetensors").unwrap();
        assert!(pricer.is_deterministic());
        
        let state = AgenticState::new();
        let tx = Transaction::new(
            AgentId::genesis(),
            AgentId::genesis(),
            TransactionContext {
                operation: Operation::Transfer { amount: 100.0 },
                energy_budget: 1.0,
                priority: Priority::Normal,
                payload: vec![],
                oracle_refs: vec![],
            },
        );
        
        // Same transaction should always get same price
        let (price1, _) = pricer.price_transaction(&tx, &state).unwrap();
        let (price2, _) = pricer.price_transaction(&tx, &state).unwrap();
        assert_eq!(price1, price2);
    }
}

File: waifu_layer/llm_pricer.py
"""WAIFU L1 - The Oracle of Truth.

LLM-based pricing engine that replaces AMMs and order books.

NOTE: despite the name, this is not real model inference. `load_quantized`
never loads real weights -- it fabricates deterministic dummy tensors, and
`model_hash` is just blake3(model_path), not a hash of actual weights.
`price_transaction` hand-encodes a feature vector and runs two matmuls
against those dummy tensors, i.e. this is dense linear algebra dressed up
as "LLM inference," faithfully preserved from the Rust original (which used
candle purely as a tensor-math library the same way). numpy is a faithful
substitute for candle here for exactly that reason.
"""

from __future__ import annotations

import math
from concurrent.futures import ThreadPoolExecutor

import blake3
import numpy as np

from .state import AgenticState
from .types import AgentId, Priority, Transaction, TransactionContext


class PricerError(Exception):
    pass


class LlmPricer:
    """The deterministic LLM pricing engine. Temperature = 0.0 for consensus-compatible inference."""

    def __init__(self, device_note: str, model_hash: bytes, embed_dim: int,
                 context_weights: np.ndarray, operation_matrices: list[np.ndarray]):
        self.device_note = device_note
        self.model_hash = model_hash
        self.embed_dim = embed_dim
        self.context_weights = context_weights
        self.operation_matrices = operation_matrices
        self.temperature = 0.0  # CRITICAL: Must be 0 for consensus

    @staticmethod
    def load_quantized(model_path: str) -> "LlmPricer":
        """Load a "quantized model" from safetensors format.

        In production, this loads actual quantized transformer weights. For
        now, initializes with deterministic pseudo-random weights.
        """
        embed_dim = 768

        model_hash = blake3.blake3(model_path.encode()).digest()

        context_weights = np.ones((embed_dim, 64), dtype=np.float32)

        operation_matrices = [
            (np.ones((embed_dim, embed_dim), dtype=np.float32) * (1.0 / (i + 1.0)))
            for i in range(8)
        ]

        return LlmPricer(
            device_note="cpu",
            model_hash=model_hash,
            embed_dim=embed_dim,
            context_weights=context_weights,
            operation_matrices=operation_matrices,
        )

    def price_transaction(self, tx: Transaction, state: AgenticState) -> tuple[float, str]:
        """Price a transaction based on its context and current network state.

        Returns (clearing_rate, pricing_rationale).
        """
        # 1. Encode the transaction context into a tensor
        context_embedding = self._encode_context(tx.context)

        # 2. Get the operation-specific pricing matrix
        op_matrix = self._get_operation_matrix(tx.context.operation)

        # 3. Query network state for supply/demand signals
        network_signal = self._compute_network_signal(state, tx.from_, tx.to)

        # 4. Perform deterministic matrix multiplication (THE CORE PRICING)
        price_vector = context_embedding @ op_matrix @ self.context_weights

        # 5. Reduce to scalar clearing rate
        raw_price = float(price_vector.sum())

        # 6. Apply network signal modulation
        clearing_rate = self._normalize_price(raw_price, network_signal, tx.context)

        # 7. Generate pricing rationale (deterministic)
        rationale = self._generate_rationale(tx.context, clearing_rate, network_signal)

        return clearing_rate, rationale

    def _encode_context(self, context: TransactionContext) -> np.ndarray:
        """Encode transaction context into embedding space."""
        features = np.zeros(self.embed_dim, dtype=np.float32)

        features[0] = context.energy_budget

        priority_idx = {
            Priority.LOW: 1,
            Priority.NORMAL: 2,
            Priority.HIGH: 3,
            Priority.CRITICAL: 4,
        }[context.priority]
        features[priority_idx] = 1.0

        op = context.operation
        if op.kind == "Transfer":
            features[10] = op.amount
            op_idx = 10
        elif op.kind == "Execute":
            op_idx = 20
        elif op.kind == "Deploy":
            features[30] = op.initial_energy
            op_idx = 30
        elif op.kind == "Bridge":
            op_idx = 40
        elif op.kind == "Stake":
            features[50] = op.amount
            op_idx = 50
        elif op.kind == "Infer":
            features[60] = op.max_tokens
            op_idx = 60
        elif op.kind == "Swap":
            op_idx = 70
        else:
            raise PricerError(f"unknown operation kind: {op.kind}")
        features[op_idx] = 1.0

        features[100] = len(context.payload) / 1024.0
        features[101] = len(context.oracle_refs)

        return features.reshape(1, self.embed_dim)

    def _get_operation_matrix(self, operation) -> np.ndarray:
        """Get the appropriate pricing matrix for an operation type."""
        idx = {
            "Transfer": 0, "Execute": 1, "Deploy": 2, "Bridge": 3,
            "Stake": 4, "Infer": 5, "Swap": 6,
        }[operation.kind]
        return self.operation_matrices[min(idx, len(self.operation_matrices) - 1)]

    def _compute_network_signal(self, state: AgenticState, from_: AgentId, to: AgentId) -> float:
        """Compute network-level supply/demand signal."""
        sender_stake = state.get_agent_stake(from_)
        receiver_stake = state.get_agent_stake(to)

        congestion = state.get_network_congestion()
        total_compute = state.get_total_compute()

        reputation_factor = math.log1p(sender_stake + receiver_stake)
        congestion_factor = 1.0 + (congestion * 2.0)
        compute_factor = math.sqrt(total_compute / 1_000_000.0)

        return reputation_factor * congestion_factor / max(compute_factor, 1.0)

    def _normalize_price(self, raw: float, network_signal: float, context: TransactionContext) -> float:
        """Normalize raw price to usable clearing rate."""
        base = abs(raw) / (self.embed_dim * 64.0)
        signaled = base * (1.0 + network_signal)

        priority_mult = {
            Priority.LOW: 0.8,
            Priority.NORMAL: 1.0,
            Priority.HIGH: 1.5,
            Priority.CRITICAL: 2.5,
        }[context.priority]

        return min(max(signaled * priority_mult, 0.0001), 1_000_000.0)

    def _generate_rationale(self, context: TransactionContext, clearing_rate: float, network_signal: float) -> str:
        """Generate deterministic pricing rationale."""
        op = context.operation
        if op.kind == "Transfer":
            op_name = f"TRANSFER({op.amount:.2f})"
        elif op.kind == "Execute":
            op_name = f"EXECUTE({op.function})"
        elif op.kind == "Deploy":
            op_name = "DEPLOY"
        elif op.kind == "Bridge":
            op_name = f"BRIDGE({op.source_chain})"
        elif op.kind == "Stake":
            op_name = f"STAKE({op.amount:.2f})"
        elif op.kind == "Infer":
            op_name = f"INFER({op.max_tokens})"
        elif op.kind == "Swap":
            op_name = "SWAP"
        else:
            op_name = op.kind

        priority_debug = {
            Priority.LOW: "Low", Priority.NORMAL: "Normal",
            Priority.HIGH: "High", Priority.CRITICAL: "Critical",
        }[context.priority]

        return (
            f"PoI-PRICE: {op_name} @ {clearing_rate:.6f} | NET_SIG: {network_signal:.4f} "
            f"| PRI: {priority_debug} | ENERGY: {context.energy_budget:.2f}"
        )

    def assess_utility_vector(self, legacy_value: float) -> float:
        """Assess the utility value of a legacy asset being bridged.

        AI utility tokens get a premium, memecoins get discounted. In
        production, this would query the agent's historical performance.
        """
        base_rate = 0.95
        scale_factor = min(math.log1p(legacy_value) / 10.0, 1.2)
        return base_rate * scale_factor

    def price_batch(self, transactions: list[Transaction], state: AgenticState) -> list[tuple[float, str] | Exception]:
        """Batch price multiple transactions for parallel execution.

        Uses a thread pool to mirror rayon::par_iter's intent -- note Python
        threads won't get true CPU parallelism for the pure-Python parts due
        to the GIL, though numpy's matmul releases the GIL during the actual
        computation, so there is real parallelism during the pricing step
        itself.
        """
        with ThreadPoolExecutor() as pool:
            def price_or_error(tx: Transaction):
                try:
                    return self.price_transaction(tx, state)
                except PricerError as e:
                    return e

            return list(pool.map(price_or_error, transactions))

    def get_model_hash(self) -> bytes:
        return self.model_hash

    def is_deterministic(self) -> bool:
        """Verify that inference is deterministic (temperature = 0)."""
        return self.temperature == 0.0
